### PR TITLE
Model for BGPv4 and route range

### PR DESCRIPTION
**Pull request by alakendu**
Please review this Model for BGPv4 and route range (IPv4 and IPv6). Please go through this test script for this model - https://github.com/open-traffic-generator/ixnetwork/blob/master/tests/test_device_bgpv4.py

### CODE REVIEW

File: api/api.yaml [hunk 1]
```diff
@@ -150,4 +150,25 @@ paths:
                 items:
                   $ref: '../result/flow.yaml#/components/schemas/Result.Flow'
 
-
+  /results/bgpinfo:
```
Comment (ajbalogh) [inline api/api.yaml:153]: change to /results/bgpv4

File: api/api.yaml [hunk 2]
```diff
@@ -150,4 +150,25 @@ paths:
...
+    description: >-
+      BGP results API
+    post:
+      tags: [Results]
+      operationId: get_bgp_infos
```
Comment (ajbalogh) [inline api/api.yaml:158]: get_bgpv4_results

File: api/api.yaml [hunk 3]
```diff
@@ -150,4 +150,25 @@ paths:
...
+      requestBody:
+        description: >-
+          The request to retrieve BGP Router statistics and learned routing information.
+        required: true
+        content:
+          application/json:
+            schema:
+              $ref: '../result/bgpinfo.yaml#/components/schemas/Result.BgpInfoRequest'
```
Comment (ajbalogh) [inline api/api.yaml:166]: Result.Bgpv4Request

File: control/control.yaml [hunk 4]
```diff
@@ -16,7 +16,7 @@ components:
       properties:
         choice:
           type: string
-          enum: [config_state, port_link_state, port_capture_state, flow_transmit_state]
+          enum: [config_state, port_link_state, port_capture_state, flow_transmit_state, bgp_contol_state]
```
Comment (ajbalogh) [inline control/control.yaml:19]: bgpv4_control_state

File: control/control.yaml [hunk 5]
```diff
@@ -25,6 +25,8 @@ components:
           $ref: '#/components/schemas/Control.PortCaptureState'
         flow_transmit_state:
           $ref: '#/components/schemas/Control.FlowTransmitState'
+        bgp_contol_state:
```
Comment (ajbalogh) [inline control/control.yaml:28]: bgpv4_control_state

File: control/control.yaml [hunk 6]
```diff
@@ -107,3 +109,23 @@ components:
...
+          type: array
+          items:
+            type: string
+            x-constraint:
+            - '/components/schemas/Device.Bgpv4'
+        state:
+          description: >-
+            The BGP Protocol state.
+          type: string
+          enum: [start, stop, pause, link_flap]
```
Comment (ajbalogh) [inline control/control.yaml:131]: remove pause

File: control/control.yaml [hunk 7]
```diff
@@ -107,3 +109,23 @@ components:
...
+      required: [state]
+      properties:
+        names:
+          description: >-
+            The names of BGP object to control.
+          type: array
+          items:
+            type: string
+            x-constraint:
+            - '/components/schemas/Device.Bgpv4'
```
Comment (ajbalogh) [inline control/control.yaml:126]: constraint is the full path of the name of the bgpv4 object -> /components/schemas/Device.Bgpv4/properties/name

File: device/bgpv4.yaml [hunk 8]
```diff
@@ -8,49 +8,75 @@ components:
...
         authentication:
           $ref: './patterns.yaml#/components/schemas/Device.Pattern'
         ttl:
           $ref: './patterns.yaml#/components/schemas/Device.Pattern'
-        dut_ipv4_address:         
+        dut_ipv4_address:
           $ref: './patterns.yaml#/components/schemas/Device.Pattern'
         ipv4:
           $ref: './ipv4.yaml#/components/schemas/Device.Ipv4'
+        bgpv4routerange:
```
Comment (ajbalogh) [inline device/bgpv4.yaml:56]: add underscores bgpv4_route_range

File: device/bgpv4.yaml [hunk 9]
```diff
@@ -8,49 +8,75 @@ components:
...
+          description: >-
+            Emulated BGPv4 route range
+          type: array
+          items:
+            $ref: '#/components/schemas/Device.Bgpv4RouteRange'
+        bgpv6routerange:
```
Comment (ajbalogh) [inline device/bgpv4.yaml:62]: add underscore bgpv6_route_range

File: device/bgpv4.yaml [hunk 10]
```diff
@@ -8,49 +8,75 @@ components:
...
 
     Device.Bgpv4RouteRange:
       description: >-
-        Emulated bgpv4 route range
+        Emulated BGPv4 route range
       type: object
       required: [name]
       properties:
         name: '../common/common.yaml#/components/x-inline/UniqueName'
+        route_count:
```
Comment (ajbalogh) [inline device/bgpv4.yaml:76]: add descriptions for all properties except inline

File: result/bgpinfo.yaml [hunk 11]
```diff
@@ -0,0 +1,56 @@
+openapi: 3.1.0
+
+info:
+  title: BGP results models
+  version: ^0.0.0
+
+components:
+  schemas:
+    Result.BgpInfoRequest:
```
Comment (ajbalogh) [inline result/bgpinfo.yaml:9]: change to Result.Bgpv4Request

File: result/bgpinfo.yaml [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+          description: >-
+            The names of BGP objects to return results for.
+            An empty list will return results for all BGP.
+          type: array
+          items:
+            type: string
+            x-constraint:
+              - "/components/schemas/Device.Bgpv4/properties/name"
+
+    Result.BgpInfo:
```
Comment (ajbalogh) [inline result/bgpinfo.yaml:24]: change to Result.Bgpv4

File: result/bgpinfo.yaml [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+      description: >-
+        BGP Router statistics and learned routing information
+      type: object
+      properties:
```
Comment (ajbalogh) [inline result/bgpinfo.yaml:28]: needs a name property

File: api/api.yaml [hunk 14]
```diff
@@ -150,4 +150,25 @@ paths:
...
+      responses:
+        '200':
+          description: >-
+            BGP Router statistics and learned routing information
+          content:
+            application/json:
+              schema:
```
Comment (ajbalogh) [inline api/api.yaml:173]: return is an array of items, see the return result for /results/port and /results/flow

File: control/control.yaml (hunk 5 above)
Comment (ajbalogh) [reply control/control.yaml:28]: remove the bgpcontrolstate for now - the /results/bgpv4 has all of that information that can easily be aggregated using pandas

File: device/bgpv4.yaml [hunk 15]
```diff
@@ -8,49 +8,75 @@ components:
...
           $ref: './patterns.yaml#/components/schemas/Device.Pattern'
-        dut_ipv4_address:         
+        dut_ipv4_address:
           $ref: './patterns.yaml#/components/schemas/Device.Pattern'
         ipv4:
           $ref: './ipv4.yaml#/components/schemas/Device.Ipv4'
+        bgpv4routerange:
+          description: >-
+            Emulated BGPv4 route range
+          type: array
```
Comment (ajbalogh) [inline device/bgpv4.yaml:59]: remove the array and make it a container node - $ref only.

File: device/bgpv4.yaml [hunk 16]
```diff
@@ -8,49 +8,75 @@ components:
...
+          items:
+            $ref: '#/components/schemas/Device.Bgpv4RouteRange'
+        bgpv6routerange:
+          description: >-
+            Emulated bgpv6 route range
+          type: array
```
Comment (ajbalogh) [inline device/bgpv4.yaml:65]: remove the array and make it a leaf node - $ref only.

**Comment by ajbalogh** [comment]
Waiting for updated commits regarding the comments I've left.

File: api/api.yaml (hunk 1 above)
Comment (alakendu) [reply api/api.yaml:153]: Agree.. Modified it accordingly.

File: api/api.yaml (hunk 2 above)
Comment (alakendu) [reply api/api.yaml:158]: Agree.. Modified it accordingly.

File: api/api.yaml (hunk 3 above)
Comment (alakendu) [reply api/api.yaml:166]: Agree.. Modified it accordingly.

File: control/control.yaml (hunk 4 above)
Comment (alakendu) [reply control/control.yaml:19]: Agree.. Modified it accordingly.

Comment (alakendu) [reply control/control.yaml:19]: I have added bgpv4_control_state to control BGP related operations. These are probable step to require this options:
1. Strat all protocols and traffic. (I think existing FlowTransmitState(state='start') should handle that part)
2. Stop BGP (I have plan to trigger that through State(BgpControlState(state='stop')))
3. Flap BGP link (trigger it  using State(BgpControlState(state='link_flap')) )

I will remove this section if this is not an actual customer use case.

File: control/control.yaml (hunk 5 above)
Comment (alakendu) [reply control/control.yaml:28]: I have added bgpv4_control_state to control BGP related operations. These are probable step to require this options:

Strat all protocols and traffic. (I think existing FlowTransmitState(state='start') should handle that part)
Stop BGP (I have plan to trigger that through State(BgpControlState(state='stop')))
Flap BGP link (trigger it using State(BgpControlState(state='link_flap')) )
I will remove this section if this is not an actual customer use case.

File: control/control.yaml (hunk 6 above)
Comment (alakendu) [reply control/control.yaml:131]: ok

File: control/control.yaml (hunk 7 above)
Comment (alakendu) [reply control/control.yaml:126]: Thanks!

File: device/bgpv4.yaml (hunk 8 above)
Comment (alakendu) [reply device/bgpv4.yaml:56]: Agree.. Modified it accordingly.

File: device/bgpv4.yaml (hunk 9 above)
Comment (alakendu) [reply device/bgpv4.yaml:62]: Agree.. Modified it accordingly.

File: result/bgpinfo.yaml (hunk 11 above)
Comment (alakendu) [reply result/bgpinfo.yaml:9]: Agree.. Modified it accordingly.

File: api/api.yaml (hunk 14 above)
Comment (alakendu) [reply api/api.yaml:173]: Thanks!

File: device/bgpv4.yaml (hunk 15 above)
Comment (alakendu) [reply device/bgpv4.yaml:59]: I am waiting for other review. Then I will restructure it accordingly.

File: result/bgpinfo.yaml (hunk 13 above)
Comment (alakendu) [reply result/bgpinfo.yaml:28]: Added it

File: result/bgpinfo.yaml (hunk 12 above)
Comment (alakendu) [reply result/bgpinfo.yaml:24]: Agree.. Modified it accordingly.

File: device/bgpv4.yaml (hunk 16 above)
Comment (alakendu) [reply device/bgpv4.yaml:65]: I am waiting for other review. Then I will restructure it accordingly.

File: control/control.yaml [hunk 17]
```diff
@@ -107,3 +109,23 @@ components:
           type: string
           enum: [start, stop]
 
+    Control.Bgpv4ControlState:
+      description: >-
+        Control BGP Protocol state to handle special case.
```
Comment (winstonliu1111) [inline control/control.yaml:114]: what does this description mean?   what is a the special case being handled?  how would a user know what that means?

File: control/control.yaml [hunk 18]
```diff
@@ -107,3 +109,23 @@ components:
...
+          type: array
+          items:
+            type: string
+            x-constraint:
+            - '/components/schemas/Device.Bgpv4/properties/name'
+        state:
+          description: >-
+            The BGP Protocol state.
+          type: string
+          enum: [start, stop, link_flap]
```
Comment (winstonliu1111) [inline control/control.yaml:131]: is link_flap the right term here for a BGP session?

File: device/bgpv4.yaml [hunk 19]
```diff
@@ -8,59 +8,134 @@ components:
...
         as_number_2_byte:
+          description: >-
+            Autonomous system (AS) number of 2 byte
           $ref: './patterns.yaml#/components/schemas/Device.Pattern'
         dut_as_number_2_byte:
+          description: >-
+            Autonomous system (AS) number of the BGP peer router (DUT)
           $ref: './patterns.yaml#/components/schemas/Device.Pattern'
         as_number_4_byte:
+          description: >-
```
Comment (winstonliu1111) [inline device/bgpv4.yaml:25]: what happens when a user specifies as_number_2_bytes and as_number_4_bytes both?  

what happens to the DUT using 4 byte as number?

File: device/bgpv4.yaml [hunk 20]
```diff
@@ -8,59 +8,134 @@ components:
...
         graceful_restart:
-          $ref: './patterns.yaml#/components/schemas/Device.Pattern'
+          description: >-
+            Graceful Restart allows the BGP forwarding state to be preserved during a BGP restart
+          type: boolean
         authentication:
           $ref: './patterns.yaml#/components/schemas/Device.Pattern'
         ttl:
+          description: >-
+            The limited number of iterations that a unit of data can experience before the data is discarded.
```
Comment (winstonliu1111) [inline device/bgpv4.yaml:54]: can experience what? before the data is discarded?

**Review by winstonliu1111 — COMMENTED** [review]
in bgp4.yaml, there are a lot of properties all referencing Device.Pattern?  just wanna make sure that this is right?    for example, for dut_ipv4_address, why would this refer to Device.Pattern when the next field ipv4 is referencing Device.Ipv4?

File: device/bgpv4.yaml (hunk 10 above)
Comment (alakendu) [reply device/bgpv4.yaml:82]: I have added them according to my understanding. I feel we need to think actual representation of aigp_metric and atomic_aggregate within Bgpv4/v6RouteRange

File: device/bgpv4.yaml [hunk 21]
```diff
@@ -8,59 +8,134 @@ components:
...
 
     Device.Bgpv4RouteRange:
       description: >-
-        Emulated bgpv4 route range
+        Emulated BGPv4 route range
       type: object
       required: [name]
       properties:
         name: '../common/common.yaml#/components/x-inline/UniqueName'
+        route_count:
```
Comment (winstonliu1111) [inline device/bgpv4.yaml:82]: @ajbalogh @ankur-sheth pondering if we actually need route count at all?  can't that be expressed as a pattern off of say the address itself?  i'm not sure what makes sense or would make it easier for other bgp implementations where it doesn't have bgp session range or route range defined?

**Review by SubhajitPalKeysight — CHANGES_REQUESTED** [review]
We also need to consider Routes - Action like ( withdraw or advertise )

withdraw_routes : 
announce_routes or advertise_routes 
these are the basic route operation even SONiC Comunity Test Cases.

File: control/control.yaml (hunk 17 above)
Comment (alakendu) [reply control/control.yaml]: dut_ipv4_address is a property where as ipv4 is protocol stack therefore one specify as Device.Pattern where as another as Device.Ipv4.
I have remove start, stop which already handle within FlowTransmitState. Sonic has test case which require withdraw_routes and advertise_routes. Therefore add BGP route action which fulfill those requirements.

File: control/control.yaml (hunk 18 above)
Comment (alakendu) [reply control/control.yaml]: I have removed link_flap  and added withdraw_routes and advertise_routes which cover Sonic use case.

File: device/bgpv4.yaml (hunk 19 above)
Comment (alakendu) [reply device/bgpv4.yaml:21]: At the time of discussion with protocol team it looks like modern DUTs are mainly support as_number_4_bytes. I have decide to expose only as_number and dut_as_number.

File: device/bgpv4.yaml (hunk 20 above)
Comment (alakendu) [reply device/bgpv4.yaml]: Current available test case do not require authentication and ttl. I have remove those. I will add with proper description when those will require.

**Comment by alakendu** [comment]
Sonic specific test case https://github.com/open-traffic-generator/ixnetwork/blob/master/tests/test_bgp_speaker.py

**Pull request merged**